dask/dataframe/core.py

```diff
@@ -2883,41 +2930,54 @@ def apply_concat_apply(args, chunk=None, aggregate=None, combine=None,
 
     token_key = tokenize(token or (chunk, aggregate), meta, args,
                          chunk_kwargs, aggregate_kwargs, combine_kwargs,
-                         split_every)
+                         split_every, split_out)
 
     # Chunk
     a = '{0}-chunk-{1}'.format(token or funcname(chunk), token_key)
     if len(args) == 1 and isinstance(args[0], _Frame) and not chunk_kwargs:
-        dsk = {(a, i): (chunk, key) for i, key in enumerate(args[0]._keys())}
+        dsk = {(a, 0, i, 0): (chunk, key) for i, key in enumerate(args[0]._keys())}
     else:
-        dsk = {(a, i): (apply, chunk, [(x._name, i) if isinstance(x, _Frame)
+        dsk = {(a, 0, i, 0): (apply, chunk, [(x._name, i) if isinstance(x, _Frame)
                                        else x for x in args], chunk_kwargs)
                for i in range(args[0].npartitions)}
 
+    # Split
+    if split_out and split_out > 1:
+        split_prefix = 'split-%s' % token_key
+        shard_prefix = 'shard-%s' % token_key
+        for i in range(args[0].npartitions):
+            dsk[(split_prefix, i)] = (hash_shard, (a, 0, i, 0), split_out, split_index)
+            for j in range(split_out):
+                dsk[(shard_prefix, 0, i, j)] = (getitem, (split_prefix, i), j)
+        a = shard_prefix
+    else:
+        split_out = 1
+
     # Combine
-    prefix = '{0}-combine-{1}-'.format(token or funcname(combine), token_key)
+    prefix = '{0}-combine-{1}'.format(token or funcname(combine), token_key)
     k = npartitions
-    b = a
+    b = prefix
     depth = 0
     while k > split_every:
-        b = prefix + str(depth)
-        for part_i, inds in enumerate(partition_all(split_every, range(k))):
-            conc = (_concat, [(a, i) for i in inds])
-            if combine_kwargs:
-                dsk[(b, part_i)] = (apply, combine, [conc], combine_kwargs)
-            else:
-                dsk[(b, part_i)] = (combine, conc)
+        for j in range(split_out):
+            for part_i, inds in enumerate(partition_all(split_every, range(k))):
```

Comment:
Swapping these two should be more efficient, as the `partition_all` is only done once then, not `split_out` times.